### ISSUE TITLE
Feature: Pwnd Passwords Support

### ISSUE THREAD

**Issue by yamikuronue** [SockDrawer/SockRPG#103]
Integrate with Pwnd Passwords.

**Comment by Bluenaxela** [comment]
@AccaliaDeElementia 

5 characters of the (hex encoded) SHA1 hash is by definition 20 bits of entropy and is what I'm referring to.

Directly getting the password back from the first five characters of the SHA1 hash is not directly feasible no, but your estimate of how it impacts brute force guessing is incorrect. It doesn't eliminate "600ish of the brazilliand and seven". It eliminates literally ~99.9999% of possible password guesses, or to put another way, eliminates in the ballpark of 999999 out of every 1000000 passwords. A brute forcing system could check the SHA1 of possible guesses in no time flat, before attempting to use the comparatively very very few remaining guesses to do an actual login to the site.

If a password was weak enough that [Eve or Mallory](https://en.wikipedia.org/wiki/Alice_and_Bob) could guess it in an average of _ten million_ attempted logins, well now they could guess it in an average of just 10 logins if they had those five characters from the hash.

Basically, a potential factor of one million reduction in brute forcing time is nothing to sneeze at, and it's _particularly_ a danger in the case of users with weak passwords that barely squeak by filters that try to avoid letting a user set weak passwords.

**Comment by Bluenaxela** [comment]
I agree that in the balance, it is likely worth it, but I do think the (admittedly small) possibility of a significant downside (which could have impact outside the context of the site, due to bad habits of re-use) makes it ethically important to indicate it in the UI.

It may also be a good idea to give an option to a server admin to use a local copy of the pwned passwords database.

Regarding other aspects of your suggested policy:

- The xkcd `correct horse battery staple` thing is flawed advice for coming up with passwords. Using a few words would be fine if the source of the words was true randomness with _no_ human intervention, but the way that is usually interpreted is people just picking words at a whim. Humans are not good sources of randomness. People rarely appreciate just how non-random their "random" picks they make are. The words that humans would pick are **heavily** biased statistically speaking, and multiple words would also have a strong statistical interdependence when picked in sequence. Even a human choice to reroll a properly random phrase generator, can introduce more bias than is ideal.
- As far as implications, I personally dislike the "passphrase" just as much as "password", due to the former it implying that using a phrase is desirable, and it tends to make people think of human-chosen phrases, which are not so great (see above). Besides, "password" is more widely understood as a term.
- It may be worth considering incorporating [zxcvbn](https://github.com/dropbox/zxcvbn). If you want to mess around with it, check this [demo here](https://lowe.github.io/tryzxcvbn/). It's not perfect, because it does not take into account statistical interdependence of different human-chosen words (for dramatic examples, it hugely overestimates the strength of "aliceinwonderland", "greetingsandsalutations" and "baconlettucetomato", but it also overestimates in much less obvious cases too), but it is the most comprehensive well-considered estimator of password strength I've seen out there, and it's biggest weakness could be combated by cautioning to not use titles, collections of related words, idioms, sayings, or intentionally grammatically correct phrases.

**Comment by Bluenaxela** [comment]
@AccaliaDeElementia 
Oh, and also... regarding the lockout for incorrect attempts, I think that's generally a good idea , but it may be worth giving thought to reducing the risk of it becoming a denial of service against a user.
One option is to have a a staged soft lockout. Consider the following concept:

1. After two failed attempts regardless of IP, within a 1 hour period, passing an "[I am not a robot](https://www.google.com/recaptcha/intro/v3beta.html)" check is required for all further attempts. Naturally, further attempts won't count towards any threshold unless that check is passed.
2. After a per-IP or per-session-cookie threshold of ten attempts is hit, disallow further attempts from both that IP and that session cookie holder for 10 minutes.
3. After a global threshold of fifteen attempts is hit, disallow further attempts from any IP **except** an IP that user successfully logged in from recently, for 10 minutes.
4. After a global threshold of thirty attempts is hit, disallow further attempts from any IP for 10 minutes.

The above rules may not be perfect and may need further refinement... but they are deliberately designed with progressively more aggressive measures, to stop someone **before** it has to get so aggressive the legitimate user gets locked out or is otherwise inconvenienced.
